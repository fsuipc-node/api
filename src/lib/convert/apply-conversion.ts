import { VM } from 'vm2';

import { Offset } from '@shared/offset';
import { RawOffsetValue } from '@shared/offset-values';
import { ConvertedOffsetValue } from '@shared/converted-offset-values';

export const applyConversion = (offset: Offset, rawOffsetValue: RawOffsetValue): ConvertedOffsetValue => {
  if (!offset.convert) {
    return rawOffsetValue;
  }

  if (offset.hasMapping) {
    if (offset.isMappingInvalid) {
      return 'INVALID_MAPPING_FUNCTION';
    }

    return offset.mappingFunction(rawOffsetValue);
  } else {
    // tslint:disable-next-line:no-eval
    if (offset.isInvalidConvertExpression) {
      return 'UNSUPPORTED_CONVERSION_EXPRESSION';
    }

    return new VM().run(
      replaceOffsetExpressionValue(offset, rawOffsetValue)
    );
  }
};

export const replaceOffsetExpressionValue = (offset: Offset, rawOffsetValue: RawOffsetValue): string => {
  return offset.convert.replace(
    new RegExp(/{VAL}/g),
    Array.isArray(rawOffsetValue)
      ? `${JSON.stringify(rawOffsetValue)}`
      : typeof rawOffsetValue === 'string'
        ? `'${rawOffsetValue}'`
        : rawOffsetValue.toString()
  );
};
