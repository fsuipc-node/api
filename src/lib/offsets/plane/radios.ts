import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const radios: OffsetList = {
  comFreq: new Offset({
    value: 0x34E,
    name: 'comFreq',
    category: OffsetCategory.RADIOS,
    description: 'Com frequency',
    type: Type.UInt16,
    convert: 'parseInt(`1` + ({VAL}).toString(16))',
    permission: 'rw',
  }),
  transponderFreq: new Offset({
    value: 0x354,
    name: 'transponderFreq',
    category: OffsetCategory.RADIOS,
    description: 'XPND transponder frequency',
    type: Type.UInt16,
    convert: 'parseInt(`1` + ({VAL}).toString(16))',
    permission: 'rw',
  }),
  nav12Select: new Offset({
    value: 0x374,
    name: 'nav12Select',
    category: OffsetCategory.RADIOS,
    description: 'NAV1/NAV2 select',
    type: Type.UInt16,
    permission: 'rw',
  }),
  dme12Select: new Offset({
    value: 0x378,
    name: 'dme12Select',
    category: OffsetCategory.RADIOS,
    description: 'DME1/DME2 select',
    type: Type.UInt16,
    permission: 'rw',
  }),
  navAdfActivate: new Offset({
    value: 0x388,
    name: 'navAdfActivate',
    category: OffsetCategory.RADIOS,
    description: 'NAV and ADF activate < FS2000',
    type: Type.UInt16,
    permission: 'w',
  }),
  comAtisActivate: new Offset({
    value: 0x38A,
    name: 'comAtisActivate',
    category: OffsetCategory.RADIOS,
    description: 'COM/ATIS activate < FS2000',
    type: Type.UInt16,
    permission: 'w',
  }),
  innerMarker: new Offset({
    value: 0xBAC,
    name: 'innerMarker',
    category: OffsetCategory.RADIOS,
    description: 'inner marker - activated when TRUE',
    type: Type.Int16,
    permission: 'r',
  }),
  middleMarker: new Offset({
    value: 0xBAE,
    name: 'middleMarker',
    category: OffsetCategory.RADIOS,
    description: 'middle marker - activated when TRUE',
    type: Type.Int16,
    permission: 'r',
  }),
  outerMarker: new Offset({
    value: 0xBB0,
    name: 'outerMarker',
    category: OffsetCategory.RADIOS,
    description: 'outer marker - activated when TRUE',
    type: Type.Int16,
    permission: 'r',
  }),
};
