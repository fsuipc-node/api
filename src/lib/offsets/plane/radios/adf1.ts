import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const ADF1: OffsetList = {
  adfFreq: new Offset({
    value: 0x34C,
    name: 'adfFreq',
    category: OffsetCategory.RADIOS,
    description: 'ADF frequency show as Binary Coded Decimal. The thousands digit and the fractional parts are provided in adfFreqExtended',
    type: Type.UInt16,
    convert: '+({VAL}).toString(16)',
    permission: 'rw',
  }),
  adfFreqExtended: new Offset({
    value: 0x356,
    name: 'adfFreqExtended',
    category: OffsetCategory.RADIOS,
    description: 'ADF frequency extended',
    type: Type.UInt16,
    permission: 'rw',
  }),
  adf1SignalStrength: new Offset({
    value: 0xC1C,
    name: 'adf1SignalStrength',
    category: OffsetCategory.RADIOS,
    description: 'ADF1 signal strength',
    type: Type.UInt32,
    permission: 'r',
  }),
  adf1BearingToNDB: new Offset({
    value: 0xC6A,
    name: 'adf1BearingToNDB',
    category: OffsetCategory.RADIOS,
    description: 'ADF1 relative bearing to NDB - degrees - -ve left, +ve right',
    convert: 'Math.round({VAL} * 360 / 65536)',
    type: Type.Int16,
    permission: 'r',
  }),
  adf1DialBearing: new Offset({
    value: 0xC6C,
    name: 'adf1DialBearing',
    category: OffsetCategory.RADIOS,
    description: 'ADF1 dial bearing - degrees',
    type: Type.UInt16,
    permission: 'rw',
  }),
};
