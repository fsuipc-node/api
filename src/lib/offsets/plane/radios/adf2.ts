import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const ADF2: OffsetList = {
  adf2Freq: new Offset({
    value: 0x2D4,
    name: 'adf2Freq',
    category: OffsetCategory.RADIOS,
    description: 'ADF 2 freq - Main 3 digits in BCD - FS2004',
    type: Type.UInt16,
    convert: '+({VAL}).toString(16)',
    permission: 'rw',
  }),
  adf2ExtendedFreq: new Offset({
    value: 0x2D6,
    name: 'adf2ExtendedFreq',
    category: OffsetCategory.RADIOS,
    description: 'ADF 2 extended freq - in BCD - FS2004',
    type: Type.UInt16,
    permission: 'rw',
  }),
  adf2RelBearing: new Offset({
    value: 0x2D8,
    name: 'adf2RelBearing',
    category: OffsetCategory.RADIOS,
    description: 'ADF2 Rel Bearing - FS2004',
    convert: 'Math.round({VAL} * 360 / 65536)',
    type: Type.Int16,
    permission: 'r',
  }),
  ndb2Identity: new Offset({
    value: 0x2DC,
    name: 'ndb2Identity',
    category: OffsetCategory.RADIOS,
    description: 'NDB2 identity - FS2004',
    type: Type.String,
    length: 6,
    permission: 'r',
  }),
  ndb2Name: new Offset({
    value: 0x2E2,
    name: 'ndb2Name',
    category: OffsetCategory.RADIOS,
    description: 'NDB2 name - FS2004',
    type: Type.String,
    length: 25,
    permission: 'r',
  }),
  ndb2IdentSoundSwitch: new Offset({
    value: 0x2FB,
    name: 'ndb2IdentSoundSwitch',
    category: OffsetCategory.RADIOS,
    description: 'NDB2 ident sound switch - FS2004',
    type: Type.Byte,
    convert: '!!{VAL}',
    permission: 'rw',
  }),
  adf2SignalStrength: new Offset({
    value: 0xC14,
    name: 'adf2SignalStrength',
    category: OffsetCategory.RADIOS,
    description: 'ADF2 signal strength',
    type: Type.UInt32,
    permission: 'r',
  }),
};
