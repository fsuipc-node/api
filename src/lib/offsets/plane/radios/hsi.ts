import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const HSI: OffsetList = {
  HSICDINeedlePosition: new Offset({
    value: 0x2F88,
    name: 'HSICDINeedlePosition',
    category: OffsetCategory.RADIOS,
    description: 'HSI CDI needle position - percent - 10 for VOR - 2.5 for LOC',
    type: Type.Double,
    convert: 'Math.round({VAL} * 100 / 127)',
    permission: 'r',
  }),
  HSIGSINeedlePosition: new Offset({
    value: 0x2F90,
    name: 'HSIGSINeedlePosition',
    category: OffsetCategory.RADIOS,
    description: 'HSI GSI needle position - percent - represents range from -0.7 to 0.7',
    type: Type.Double,
    convert: 'Math.round({VAL} * 100 / 119)',
    permission: 'r',
  }),
  HSISpeed: new Offset({
    value: 0x2F98,
    name: 'HSISpeed',
    category: OffsetCategory.RADIOS,
    description: 'HSI speed',
    type: Type.Double,
    convert: 'Math.round({VAL} * 1.94384)',
    permission: 'r',
  }),
  HSIDistance: new Offset({
    value: 0x2FA0,
    name: 'HSIDistance',
    category: OffsetCategory.RADIOS,
    description: 'HSI distance - meters',
    type: Type.Double,
    permission: 'r',
  }),
  HSIBearing: new Offset({
    value: 0x2FA8,
    name: 'HSIBearing',
    category: OffsetCategory.RADIOS,
    description: 'HSI bearing - degrees',
    type: Type.UInt16,
    permission: 'r',
  }),
  HSICDIValid: new Offset({
    value: 0x2FAA,
    name: 'HSICDIValid',
    category: OffsetCategory.RADIOS,
    description: 'HSI CDI valid flag',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  HSIGSIValid: new Offset({
    value: 0x2FAB,
    name: 'HSIGSIValid',
    category: OffsetCategory.RADIOS,
    description: 'HSI GSI valid flag',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  HSIBearingValid: new Offset({
    value: 0x2FAC,
    name: 'HSIBearingValid',
    category: OffsetCategory.RADIOS,
    description: 'HSI Bearing valid flag',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  HSIToFrom: new Offset({
    value: 0x2FAD,
    name: 'HSIToFrom',
    category: OffsetCategory.RADIOS,
    description: 'HSI To/From flag',
    convert: 'vorToFrom',
    mapping: true,
    type: Type.Byte,
    permission: 'r',
  }),
  HSIHasLocaliser: new Offset({
    value: 0x2FAE,
    name: 'HSIHasLocaliser',
    category: OffsetCategory.RADIOS,
    description: 'HSI has localiser flag',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  HSIIdent: new Offset({
    value: 0x2FB0,
    name: 'HSIIdent',
    category: OffsetCategory.RADIOS,
    description: 'HSI ident string',
    length: 6,
    type: Type.String,
    permission: 'r',
  }),
};
