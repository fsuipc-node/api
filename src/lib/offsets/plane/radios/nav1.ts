import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const NAV1: OffsetList = {
  vor1DMEDistance: new Offset({
    value: 0x300,
    name: 'vor1DMEDistance',
    category: OffsetCategory.RADIOS,
    description: 'VOR1 DME distance - nm',
    convert: '{VAL} / 10',
    type: Type.UInt16,
    permission: 'r',
  }),
  vor1DMESpeed: new Offset({
    value: 0x302,
    name: 'vor1DMESpeed',
    category: OffsetCategory.RADIOS,
    description: 'VOR1 DME speed - kt',
    convert: '{VAL} / 10',
    type: Type.UInt16,
    permission: 'r',
  }),
  vor1DMETimeToStation: new Offset({
    value: 0x304,
    name: 'vor1DMETimeToStation',
    category: OffsetCategory.RADIOS,
    description: 'VOR1 DME time to station - sec',
    convert: '{VAL} / 10',
    type: Type.UInt16,
    permission: 'r',
  }),
  nav1Freq: new Offset({
    value: 0x350,
    name: 'nav1Freq',
    category: OffsetCategory.RADIOS,
    description: 'NAV1 frequency',
    type: Type.UInt16,
    convert: 'parseInt(`1` + ({VAL}).toString(16))',
    permission: 'rw',
  }),
  hasNav1: new Offset({
    value: 0x7A0,
    name: 'hasNav1',
    category: OffsetCategory.RADIOS,
    description: 'has NAV1',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  vor1OrILSGlideSlopeLatitude: new Offset({
    value: 0x85C,
    name: 'vor1OrILSGlideSlopeLatitude',
    category: OffsetCategory.RADIOS,
    description: 'VOR 1 latitude OR NAV1 ILS glideslope latitude',
    type: Type.Int32,
    convert: '{VAL} * 90 / 10001750',
    permission: 'r',
  }),
  vor1OrILSGlideSlopeLongitude: new Offset({
    value: 0x864,
    name: 'vor1OrILSGlideSlopeLongitude',
    category: OffsetCategory.RADIOS,
    description: 'VOR 1 or NAV 1 ILS glideslope longitude',
    type: Type.Int32,
    convert: '{VAL} * 360 / (65536 * 65536)',
    permission: 'r',
  }),
  vor1OrILSGlideSlopeElevation: new Offset({
    value: 0x86C,
    name: 'vor1OrILSGlideSlopeElevation',
    category: OffsetCategory.RADIOS,
    description: 'VOR 1 or NAV1 ILS glideslope elevation - meters - FS2002+',
    type: Type.Int32,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
  vor1ILSLocHeadingTrue: new Offset({
    value: 0x870,
    name: 'vor1LocHeadingTrue',
    category: OffsetCategory.RADIOS,
    description: 'VOR 1 ILS LOC heading - TRUE - -180 different to aircraft direction to follow localiser',
    type: Type.UInt16,
    convert: '+({VAL} * 360 / 65536).toFixed(2)',
    permission: 'r',
  }),
  vor1GlideSlopeAngle: new Offset({
    value: 0x872,
    name: 'vor1GlideSlopeAngle',
    category: OffsetCategory.RADIOS,
    description: 'VOR 1 ILS glideslope Angle',
    type: Type.Int16,
    convert: '+({VAL} * 360 / 65536).toFixed(2)',
    permission: 'r',
  }),
  vor1LocLatitude: new Offset({
    value: 0x874,
    name: 'vor1LocLatitude',
    category: OffsetCategory.RADIOS,
    description: 'VOR 1 or NAV1 ILS LOC latitude - FS2002+',
    type: Type.Int32,
    convert: '{VAL} * 90 / 10001750',
    permission: 'r',
  }),
  vor1LocLongitude: new Offset({
    value: 0x878,
    name: 'vor1LocLongitude',
    category: OffsetCategory.RADIOS,
    description: 'VOR 1 or NAV1 ILS LOC longitude - FS2002+',
    type: Type.Int32,
    convert: '{VAL} * 360 / (65536 * 65536)',
    permission: 'r',
  }),
  vor1LocElevation: new Offset({
    value: 0x87C,
    name: 'vor1LocElevation',
    category: OffsetCategory.RADIOS,
    description: 'VOR 1 or NAV1 ILS LOC elevation - meters - FS2002+',
    type: Type.Int32,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
  vor1DmeLatitude: new Offset({
    value: 0x880,
    name: 'vor1DmeLatitude',
    category: OffsetCategory.RADIOS,
    description: 'VOR 1 DME latitude - FS2002+',
    type: Type.Int32,
    convert: '{VAL} * 90 / 10001750',
    permission: 'r',
  }),
  vor1DmeLongitude: new Offset({
    value: 0x884,
    name: 'vor1DmeLongitude',
    category: OffsetCategory.RADIOS,
    description: 'VOR 1 DME longitude - FS2002+',
    type: Type.Int32,
    convert: '{VAL} * 360 / (65536 * 65536)',
    permission: 'r',
  }),
  vor1DmeElevation: new Offset({
    value: 0x88A,
    name: 'vor1DmeElevation',
    category: OffsetCategory.RADIOS,
    description: 'VOR 1 DME elevation - ft - FS2002+',
    type: Type.Int32,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
  dme1Distance: new Offset({
    value: 0xC29,
    name: 'dme1Distance',
    category: OffsetCategory.RADIOS,
    description: 'DME1 distance - nm',
    convert: '+{VAL}',
    type: Type.String,
    length: 5,
    permission: 'r',
  }),
  dme1Speed: new Offset({
    value: 0xC2E,
    name: 'dme1Speed',
    category: OffsetCategory.RADIOS,
    description: 'DME1 speed - kt',
    convert: '+{VAL}',
    type: Type.String,
    length: 5,
    permission: 'r',
  }),
  nav1MagVar: new Offset({
    value: 0xC40,
    name: 'nav1MagVar',
    category: OffsetCategory.RADIOS,
    description: 'NAV1 mag var',
    convert: 'Math.round({VAL} * 360 / 65536)',
    type: Type.Int16,
    permission: 'r',
  }),
  nav1LocNeedlePosition: new Offset({
    value: 0xC48,
    name: 'nav1LocNeedlePosition',
    category: OffsetCategory.RADIOS,
    description: 'NAV1 loc needle position - percent - -ve left +ve right',
    convert: 'Math.round({VAL} / 127 * 100)',
    type: Type.SByte,
    permission: 'r',
  }),
  nav1GlideslopeNeedlePosition: new Offset({
    value: 0xC49,
    name: 'nav1GlideslopeNeedlePosition',
    category: OffsetCategory.RADIOS,
    description: 'NAV1 glideslope needle position - percent - -ve up +ve down',
    convert: 'Math.round({VAL} / 127 * 100)',
    type: Type.SByte,
    permission: 'r',
  }),
  nav1BackCourseFlags: new Offset({
    value: 0xC4A,
    name: 'nav1BackCourseFlags',
    category: OffsetCategory.RADIOS,
    description: 'NAV1 glideslope flags',
    convert: 'navBackCourseFlags',
    mapping: true,
    type: Type.ByteArray,
    length: 8,
    permission: 'r',
  }),
  nav1ToFrom: new Offset({
    value: 0xC4B,
    name: 'nav1ToFrom',
    category: OffsetCategory.RADIOS,
    description: 'NAV1 TO/FROM flag - values: off/to/from',
    convert: 'vorToFrom',
    mapping: true,
    type: Type.Byte,
    permission: 'r',
  }),
  nav1GlideslopeActive: new Offset({
    value: 0xC4C,
    name: 'nav1GlideslopeActive',
    category: OffsetCategory.RADIOS,
    description: 'NAV1 glideslope flag',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  nav1Capabilities: new Offset({
    value: 0xC4D,
    name: 'nav1Capabilities',
    category: OffsetCategory.RADIOS,
    description: 'NAV1 capabilities',
    convert: 'navCapabilities',
    mapping: true,
    type: Type.ByteArray,
    length: 8,
    permission: 'r',
  }),
  nav1Obs: new Offset({
    value: 0xC4E,
    name: 'nav1Obs',
    category: OffsetCategory.RADIOS,
    description: 'NAV1 obs setting - degrees',
    type: Type.Int16,
    permission: 'rw',
  }),
  nav1Radial: new Offset({
    value: 0xC50,
    name: 'nav1Radial',
    category: OffsetCategory.RADIOS,
    description: 'NAV1 radial - degrees - MAG for VOR, TRUE for ILS LOC',
    convert: 'Math.round({VAL} * 360 / 65536)',
    type: Type.UInt16,
    permission: 'r',
  }),
  nav1SignalStrength: new Offset({
    value: 0xC52,
    name: 'nav1SignalStrength',
    category: OffsetCategory.RADIOS,
    description: 'NAV1 signal strength',
    type: Type.UInt32,
    permission: 'r',
  }),
  nav1BearingToVOR: new Offset({
    value: 0xC56,
    name: 'nav1BearingToVOR',
    category: OffsetCategory.RADIOS,
    description: 'NAV1 relative bearing to VOR - degrees',
    type: Type.UInt16,
    permission: 'r',
  }),
};
