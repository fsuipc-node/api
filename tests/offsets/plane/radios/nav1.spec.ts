import { NAV1 as offsets } from '@offsets/plane/radios/nav1';
import { replaceOffsetExpressionValue } from '@convert/apply-conversion';

describe('offset - plane/radios/nav1', () => {
  const offsetsTestCases = [
    { name: 'vor1DMEDistance', value: 250, expectedResult: 25 },
    { name: 'vor1DMESpeed', value: 350, expectedResult: 35 },
    { name: 'vor1DMETimeToStation', value: 450, expectedResult: 45 },
    { name: 'nav1Freq', value: 13584, expectedResult: 13510 },
    { name: 'hasNav1', value: 0, expectedResult: false },
    { name: 'hasNav1', value: 1, expectedResult: true },
    { name: 'vor1OrILSGlideSlopeLatitude', value: 5198687, expectedResult: 46.77999650061239 },
    { name: 'vor1OrILSGlideSlopeLongitude', value: -851715875, expectedResult: -71.38999993912876 },
    { name: 'vor1OrILSGlideSlopeElevation', value: 92, expectedResult: 301.84 },
    { name: 'vor1ILSLocHeadingTrue', value: 27346, expectedResult: 150.22 },
    { name: 'vor1GlideSlopeAngle', value: 27346, expectedResult: 150.22 },
    { name: 'vor1LocLatitude', value: 5198687, expectedResult: 46.77999650061239 },
    { name: 'vor1LocLongitude', value: -851715875, expectedResult: -71.38999993912876 },
    { name: 'vor1LocElevation', value: 92, expectedResult: 301.84 },
    { name: 'vor1DmeLatitude', value: 5198687, expectedResult: 46.77999650061239 },
    { name: 'vor1DmeLongitude', value: -851715875, expectedResult: -71.38999993912876 },
    { name: 'vor1DmeElevation', value: 92, expectedResult: 301.84 },
    { name: 'dme1Distance', value: '92.5', expectedResult: 92.5 },
    { name: 'dme1Distance', value: '92.5 ', expectedResult: 92.5 },
    { name: 'dme1Distance', value: '92.50', expectedResult: 92.5 },
    { name: 'dme1Distance', value: '112.5', expectedResult: 112.5 },
    { name: 'dme1Distance', value: '112.0', expectedResult: 112 },
    { name: 'dme1Distance', value: '112. ', expectedResult: 112 },
    { name: 'dme1Speed', value: '92.5', expectedResult: 92.5 },
    { name: 'dme1Speed', value: '92.5 ', expectedResult: 92.5 },
    { name: 'dme1Speed', value: '92.50', expectedResult: 92.5 },
    { name: 'dme1Speed', value: '112.5', expectedResult: 112.5 },
    { name: 'dme1Speed', value: '112.0', expectedResult: 112 },
    { name: 'dme1Speed', value: '112. ', expectedResult: 112 },
    { name: 'nav1MagVar', value: 9102, expectedResult: 50 },
    { name: 'nav1LocNeedlePosition', value: 127, expectedResult: 100 },
    { name: 'nav1LocNeedlePosition', value: -127, expectedResult: -100 },
    { name: 'nav1GlideslopeNeedlePosition', value: 127, expectedResult: 100 },
    { name: 'nav1GlideslopeNeedlePosition', value: -127, expectedResult: -100 },
    { name: 'nav1GlideslopeActive', value: 0, expectedResult: false },
    { name: 'nav1GlideslopeActive', value: 1, expectedResult: true },
    { name: 'nav1Radial', value: 9102, expectedResult: 50 },
  ];

  describe('offsets list', () => {
    it('should have required properties', () => {
      expect(offsets).toMatchSnapshot();
    });
  });

  offsetsTestCases.forEach(testedOffset => {
    describe(testedOffset.name, () => {
      it('should convert data properly', () => {
        const convertExpression = replaceOffsetExpressionValue(offsets[testedOffset.name], testedOffset.value);

        // tslint:disable-next-line:no-eval
        expect(eval(convertExpression)).toEqual(testedOffset.expectedResult);
      });
    });
  });
});
