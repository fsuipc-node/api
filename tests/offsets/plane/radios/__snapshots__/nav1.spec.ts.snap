// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`offset - plane/radios/nav1 offsets list should have required properties 1`] = `
Object {
  "dme1Distance": Offset {
    "category": "radios",
    "convert": "+{VAL}",
    "description": "DME1 distance - nm",
    "length": 5,
    "mapping": undefined,
    "name": "dme1Distance",
    "permission": "r",
    "type": 11,
    "value": 3113,
  },
  "dme1Speed": Offset {
    "category": "radios",
    "convert": "+{VAL}",
    "description": "DME1 speed - kt",
    "length": 5,
    "mapping": undefined,
    "name": "dme1Speed",
    "permission": "r",
    "type": 11,
    "value": 3118,
  },
  "hasNav1": Offset {
    "category": "radios",
    "convert": "!!{VAL}",
    "description": "has NAV1",
    "length": undefined,
    "mapping": undefined,
    "name": "hasNav1",
    "permission": "rw",
    "type": 6,
    "value": 1952,
  },
  "nav1BackCourseFlags": Offset {
    "category": "radios",
    "convert": "navBackCourseFlags",
    "description": "NAV1 glideslope flags",
    "length": 8,
    "mapping": true,
    "name": "nav1BackCourseFlags",
    "permission": "r",
    "type": 10,
    "value": 3146,
  },
  "nav1BearingToVOR": Offset {
    "category": "radios",
    "convert": undefined,
    "description": "NAV1 relative bearing to VOR - degrees",
    "length": undefined,
    "mapping": undefined,
    "name": "nav1BearingToVOR",
    "permission": "r",
    "type": 5,
    "value": 3158,
  },
  "nav1Capabilities": Offset {
    "category": "radios",
    "convert": "navCapabilities",
    "description": "NAV1 capabilities",
    "length": 8,
    "mapping": true,
    "name": "nav1Capabilities",
    "permission": "r",
    "type": 10,
    "value": 3149,
  },
  "nav1Freq": Offset {
    "category": "radios",
    "convert": "parseInt(\`1\` + ({VAL}).toString(16))",
    "description": "NAV1 frequency",
    "length": undefined,
    "mapping": undefined,
    "name": "nav1Freq",
    "permission": "rw",
    "type": 5,
    "value": 848,
  },
  "nav1GlideslopeActive": Offset {
    "category": "radios",
    "convert": "!!{VAL}",
    "description": "NAV1 glideslope flag",
    "length": undefined,
    "mapping": undefined,
    "name": "nav1GlideslopeActive",
    "permission": "r",
    "type": 0,
    "value": 3148,
  },
  "nav1GlideslopeNeedlePosition": Offset {
    "category": "radios",
    "convert": "Math.round({VAL} / 127 * 100)",
    "description": "NAV1 glideslope needle position - percent - -ve up +ve down",
    "length": undefined,
    "mapping": undefined,
    "name": "nav1GlideslopeNeedlePosition",
    "permission": "r",
    "type": 1,
    "value": 3145,
  },
  "nav1LocNeedlePosition": Offset {
    "category": "radios",
    "convert": "Math.round({VAL} / 127 * 100)",
    "description": "NAV1 loc needle position - percent - -ve left +ve right",
    "length": undefined,
    "mapping": undefined,
    "name": "nav1LocNeedlePosition",
    "permission": "r",
    "type": 1,
    "value": 3144,
  },
  "nav1MagVar": Offset {
    "category": "radios",
    "convert": "Math.round({VAL} * 360 / 65536)",
    "description": "NAV1 mag var",
    "length": undefined,
    "mapping": undefined,
    "name": "nav1MagVar",
    "permission": "r",
    "type": 2,
    "value": 3136,
  },
  "nav1Obs": Offset {
    "category": "radios",
    "convert": undefined,
    "description": "NAV1 obs setting - degrees",
    "length": undefined,
    "mapping": undefined,
    "name": "nav1Obs",
    "permission": "rw",
    "type": 2,
    "value": 3150,
  },
  "nav1Radial": Offset {
    "category": "radios",
    "convert": "Math.round({VAL} * 360 / 65536)",
    "description": "NAV1 radial - degrees - MAG for VOR, TRUE for ILS LOC",
    "length": undefined,
    "mapping": undefined,
    "name": "nav1Radial",
    "permission": "r",
    "type": 5,
    "value": 3152,
  },
  "nav1SignalStrength": Offset {
    "category": "radios",
    "convert": undefined,
    "description": "NAV1 signal strength",
    "length": undefined,
    "mapping": undefined,
    "name": "nav1SignalStrength",
    "permission": "r",
    "type": 6,
    "value": 3154,
  },
  "nav1ToFrom": Offset {
    "category": "radios",
    "convert": "vorToFrom",
    "description": "NAV1 TO/FROM flag - values: off/to/from",
    "length": undefined,
    "mapping": true,
    "name": "nav1ToFrom",
    "permission": "r",
    "type": 0,
    "value": 3147,
  },
  "vor1DMEDistance": Offset {
    "category": "radios",
    "convert": "{VAL} / 10",
    "description": "VOR1 DME distance - nm",
    "length": undefined,
    "mapping": undefined,
    "name": "vor1DMEDistance",
    "permission": "r",
    "type": 5,
    "value": 768,
  },
  "vor1DMESpeed": Offset {
    "category": "radios",
    "convert": "{VAL} / 10",
    "description": "VOR1 DME speed - kt",
    "length": undefined,
    "mapping": undefined,
    "name": "vor1DMESpeed",
    "permission": "r",
    "type": 5,
    "value": 770,
  },
  "vor1DMETimeToStation": Offset {
    "category": "radios",
    "convert": "{VAL} / 10",
    "description": "VOR1 DME time to station - sec",
    "length": undefined,
    "mapping": undefined,
    "name": "vor1DMETimeToStation",
    "permission": "r",
    "type": 5,
    "value": 772,
  },
  "vor1DmeElevation": Offset {
    "category": "radios",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "VOR 1 DME elevation - ft - FS2002+",
    "length": undefined,
    "mapping": undefined,
    "name": "vor1DmeElevation",
    "permission": "r",
    "type": 3,
    "value": 2186,
  },
  "vor1DmeLatitude": Offset {
    "category": "radios",
    "convert": "{VAL} * 90 / 10001750",
    "description": "VOR 1 DME latitude - FS2002+",
    "length": undefined,
    "mapping": undefined,
    "name": "vor1DmeLatitude",
    "permission": "r",
    "type": 3,
    "value": 2176,
  },
  "vor1DmeLongitude": Offset {
    "category": "radios",
    "convert": "{VAL} * 360 / (65536 * 65536)",
    "description": "VOR 1 DME longitude - FS2002+",
    "length": undefined,
    "mapping": undefined,
    "name": "vor1DmeLongitude",
    "permission": "r",
    "type": 3,
    "value": 2180,
  },
  "vor1GlideSlopeAngle": Offset {
    "category": "radios",
    "convert": "+({VAL} * 360 / 65536).toFixed(2)",
    "description": "VOR 1 ILS glideslope Angle",
    "length": undefined,
    "mapping": undefined,
    "name": "vor1GlideSlopeAngle",
    "permission": "r",
    "type": 2,
    "value": 2162,
  },
  "vor1ILSLocHeadingTrue": Offset {
    "category": "radios",
    "convert": "+({VAL} * 360 / 65536).toFixed(2)",
    "description": "VOR 1 ILS LOC heading - TRUE - -180 different to aircraft direction to follow localiser",
    "length": undefined,
    "mapping": undefined,
    "name": "vor1LocHeadingTrue",
    "permission": "r",
    "type": 5,
    "value": 2160,
  },
  "vor1LocElevation": Offset {
    "category": "radios",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "VOR 1 or NAV1 ILS LOC elevation - meters - FS2002+",
    "length": undefined,
    "mapping": undefined,
    "name": "vor1LocElevation",
    "permission": "r",
    "type": 3,
    "value": 2172,
  },
  "vor1LocLatitude": Offset {
    "category": "radios",
    "convert": "{VAL} * 90 / 10001750",
    "description": "VOR 1 or NAV1 ILS LOC latitude - FS2002+",
    "length": undefined,
    "mapping": undefined,
    "name": "vor1LocLatitude",
    "permission": "r",
    "type": 3,
    "value": 2164,
  },
  "vor1LocLongitude": Offset {
    "category": "radios",
    "convert": "{VAL} * 360 / (65536 * 65536)",
    "description": "VOR 1 or NAV1 ILS LOC longitude - FS2002+",
    "length": undefined,
    "mapping": undefined,
    "name": "vor1LocLongitude",
    "permission": "r",
    "type": 3,
    "value": 2168,
  },
  "vor1OrILSGlideSlopeElevation": Offset {
    "category": "radios",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "VOR 1 or NAV1 ILS glideslope elevation - meters - FS2002+",
    "length": undefined,
    "mapping": undefined,
    "name": "vor1OrILSGlideSlopeElevation",
    "permission": "r",
    "type": 3,
    "value": 2156,
  },
  "vor1OrILSGlideSlopeLatitude": Offset {
    "category": "radios",
    "convert": "{VAL} * 90 / 10001750",
    "description": "VOR 1 latitude OR NAV1 ILS glideslope latitude",
    "length": undefined,
    "mapping": undefined,
    "name": "vor1OrILSGlideSlopeLatitude",
    "permission": "r",
    "type": 3,
    "value": 2140,
  },
  "vor1OrILSGlideSlopeLongitude": Offset {
    "category": "radios",
    "convert": "{VAL} * 360 / (65536 * 65536)",
    "description": "VOR 1 or NAV 1 ILS glideslope longitude",
    "length": undefined,
    "mapping": undefined,
    "name": "vor1OrILSGlideSlopeLongitude",
    "permission": "r",
    "type": 3,
    "value": 2148,
  },
}
`;
